[CMS] Speed up the cache-busting solution of build

ncksllvn opened the pull request:
## Description
This pull request modifies our cache-busting solution for Webpack-generated JS/CSS. Currently, we apply a generated hash to the file names of Webpack JS/CSS. Then, we cycle through our HTML files and update non-hashed references to those assets via regex. This PR changes that strategy to instead use the DOM by adding `data-entry-name` properties into our DOM references to Webpack assets, then just searching for those nodes from our build. This reduces this build step for vagovprod builds from `69824ms` (from latest master build) to `169ms` (this PR build).

## Testing done
- Local builds of vagovprod
    - Inspecting HTML output and file names
    - E2E tests and broken links passing means that nothing can be totally broken
- Local testing of TeamSite
- Local testing of Preview server

## Screenshots
![image](https://user-images.githubusercontent.com/1915775/67403509-0a5d0c80-f580-11e9-8f50-4d11d2161660.png)



## Acceptance criteria
- [ ] Cache-busting step of build is faster and doesn't break the website

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

Commit:
[CMS] Speed up the cache-busting solution of build

Commit:
Fix for TeamSite-required assets

ncksllvn commented on src/site/includes/header.html:
```diff
@@ -33,27 +34,31 @@
+
+<script defer
+  data-entry-name="{{ entryname | default: 'static-pages' }}.js"
+  src="/generated/{{ entryname | default: 'static-pages' }}.entry.js"></script>
```
This PR adds that `data-entry-name` attribute to our Webpack references. This way, we can quickly filter our DOM in our `addAssetHashes` build step.

ncksllvn replied:
The values of `data-entry-names` align exactly with the keys in the `generated/file-manifest.json`.

ElijahLynn replied:
These seem like good inline comments.

cvalarida replied:
+1 for documentation somehow, but I'm not sure if inline comments are the best choice (annoyingly). I'm pretty sure they'll show up in production, which I think we probably don't want.

ncksllvn replied:
I renamed the plugin to `process-entry-names` and changed it so that we no longer need `src` or `href` values at all for Webpack assets. I think this should make it clearer that the `data-entry-name` attribute is meaningful.

ncksllvn replied:
I really don't think there's much value in a Liquid comment, but I can do that if you feel strongly.

ncksllvn commented on src/site/stages/build/index.js:
```diff
@@ -231,6 +232,10 @@ function defaultBuild(BUILD_OPTIONS) {
+  smith.use(
+    addAssetHashes(BUILD_OPTIONS),
+    'Add hashes to Webpack assets for browser cache-busting',
+  );
```
I had to move this out of `configureAssets` to access the `file.dom` property.

ncksllvn commented on src/site/stages/build/plugins/add-asset-hashes.js:
```diff
@@ -1,69 +1,85 @@
+
+    files[teamSitePath] = file;
+  }
+}
```
I hate this TeamSite stuff, but it's necessary. View the source for https://benefits.va.gov/ to see the hardcoded references to these assets.

ncksllvn commented on src/site/stages/build/plugins/configure-assets.js:
```diff
@@ -24,19 +19,13 @@ function configureAssets(smith, buildOptions) {
+      webpackMetalsmithConnect.compileAssets(buildOptions),
+      'Build Webpack assets',
+    );
   }
```
This just moved the `addAssetHashes` plugin to `build/index.js`.

ncksllvn commented on src/site/stages/build/plugins/add-asset-hashes.js:
```diff
@@ -1,69 +1,85 @@
+  if (isDevBuild) {
+    const noop = () => {};
+    return noop;
+  }
```
No hashes on local/dev.

cvalarida commented on src/site/stages/build/plugins/add-asset-hashes.js:
```diff
@@ -1,69 +1,85 @@
-        const newKey = `${matchData[1]}.entry.js`;
-        manifest[newKey] = originalManifest[originalManifestKey];
+  for (const fileManifestKey of Object.keys(teamsiteInjectionBundles)) {
+    const teamSitePath = teamsiteInjectionBundles[fileManifestKey];
```
Alternatively, you could use
```js
for (const [fileManifestKey, teamSitePath] of Object.entries(teamsiteInjectionBuntles)) {
  ...
}
```

ncksllvn replied:
Updated. Prettier made it nice and ugly for us 🤷‍♂

cvalarida commented on src/site/stages/build/plugins/add-asset-hashes.js:
```diff
@@ -1,69 +1,85 @@
+  // TeamSite pages such as "benefits.va.gov" have hardcoded references to certain
+  // JavaScript bundles so that we can inject our header/footer into pages outside
+  // of our source. Therefore, we can't apply a hash to those bundles.
+  const teamsiteInjectionBundles = {
```
What do you think of renaming this to `teamSiteInjectionBundles` (capital `S`) or renaming `teamSitePath` with a lowercase `s` (`teamsitePath`) for consistency?

ncksllvn replied:
Fine with me.

cvalarida commented on src/site/stages/build/plugins/add-asset-hashes.js:
```diff
@@ -1,69 +1,85 @@
+    if (!hashedFileName) {
+      if (buildOptions.entry) {
+        // A command arg was passed to build only certain areas of the site, so
+        // it makes that it would be missing.
```
This would mean we have a production or staging build with an `--entry` flag.
:thinking: That's weird, but probably alright.

ncksllvn replied:
No, I think `--entry` is only used for localhost builds.  I don't know why this would happen. I'm just going by the comment that was there on `addAssetHashes`.

cvalarida commented on src/site/stages/build/plugins/add-asset-hashes.js:
```diff
@@ -1,69 +1,85 @@
-    const originalManifest = JSON.parse(files[manifestKey].contents.toString());
+const FILE_MANIFEST_FILENAME = 'generated/file-manifest.json';
+
+function copyAssetsToTeamSitePaths(buildOptions, files, entryNamesDictionary) {
```
How can we test this?

ncksllvn replied:
I tested it by looking at the output of `build/localhost/generated/` and confirming the non-hashed bundles are there. As far as an integration test, probably just by pinging the assets location and confirming the paths exist, or by going all-out and writing a suite for TeamSite E2E tests since we don't have those now.

cvalarida reviewed (commented):
Looks great! I just want to make sure we have some way to verify the teamsite JS bundles are golden before I give it the ol' rubber stamp.

bkjohnson commented on src/site/stages/build/plugins/add-asset-hashes.js:
```diff
@@ -1,69 +1,85 @@
+    'polyfills.js': 'generated/polyfills.entry.js',
+    'styleConsolidated.css': 'generated/styleConsolidated.css',
+    'static-pages.css': 'generated/static-pages.css',
+  };
```
This may be overkill, but it would eliminate some redundancy:

```js
  const teamsiteFiles = [
    'proxy-rewrite.js',
    'vendor.js',
    'polyfills.js',
    'styleConsolidated.css',
    'static-pages.css'
  ]

  const teamsiteInjectionBundles = teamsiteFiles.reduce((obj, fileName) => { 
    const [ name, ext] = fileName.split('.')
    obj[fileName] = `generated/${name}.entry.${ext}`
    return obj
    }, {})
```

ncksllvn replied:
I personally prefer the explicitness of the hardcoded string values, because this is kind of an unclear operation for an edge of the whole build.

ncksllvn replied:
Also, `.css` doesn't get the `.entry` piece applied 🤷‍♂

Commit:
Also generate for local/dev builds

Commit:
Rename to process-entry-names

Commit:
Fixes for Preview Server

Commit:
Rename func

ncksllvn commented on src/site/stages/build/plugins/process-entry-names.js:
```diff
@@ -0,0 +1,103 @@
+  };
+}
+
+module.exports = addAssetHashes;
```
I renamed the `addAssetHashes` plugin to `processEntryNames` because -. 
- I wanted to also apply it to local/dev builds, which don't use hashes but do change `.js` to `.entry.js`
- By enabling it for local/dev, we can always generated the asset path for the `script src` or `link href` attributes and not have to spec. This way, if there's an issue with the plugin, we may spot it on local as opposed to a higher env.
- Increased visibility into what this plugin does

ncksllvn commented on src/site/stages/build/plugins/process-entry-names.js:
```diff
@@ -0,0 +1,103 @@
+  };
+}
+
+module.exports = processEntryNames;
```
I renamed the `addAssetHashes` plugin to `processEntryNames` because -. 
- I wanted to also apply it to local/dev builds, which don't use hashes but do change `.js` to `.entry.js`
- By enabling it for local/dev, we can always generated the asset path for the `script src` or `link href` attributes and not have to spec. This way, if there's an issue with the plugin, we may spot it on local as opposed to a higher env.
- Increased visibility into what this plugin does

cvalarida replied:
:100: I dig it!

Commit:
Review comments

ncksllvn commented on script/preview.js:
```diff
@@ -97,6 +97,7 @@ app.get('/preview', async (req, res, next) => {
   try {
     const smith = await createPipieline({
       ...options,
+      isPreviewServer: true,
```
We need a flag to send to `process-entry-names` to indicate that we should not try to do that whole TeamSite step. This was how the old plugin `addAssetHashes` worked 🤷‍♂

cvalarida reviewed (approved):
This is awesome. Thanks! It's probably out of scope for this PR since we didn't have any before, but we should seriously consider testing this whole pipeline. We'll catch if the React app bundles are messed up in some way by our e2e tests, but the teamsite bundles aren't tested at all, are they?

ncksllvn commented:
Thanks @cvalarida! You identified this as a bottleneck in the first place so props to you 😄Yeah, we need some sort of integration test for the Preview Server and the TeamSite header/footer injection. I'm gonna write the tickets up for adding tests for those. They're special cases though, so they'd take some creative thinking.

ncksllvn commented:
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/2704
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/2705

Commit:
Add JS to broken link checker

Commit:
Validate entry name from plugin

Commit:
Remove pointless if-statement

ncksllvn commented on src/site/stages/build/plugins/process-entry-names.js:
```diff
@@ -0,0 +1,108 @@
+
+        if (!buildOptions.watch && !entryExists) {
+          throw new Error(`Entry Name "${entryName}" was not found.`);
+        }
```
I realized we can check that the bundle actually exists and then throw an error if it doesn't. This doesn't work in watch-mode since the file is only in Webpack memory, but it does for standard builds.

ncksllvn commented:
@cvalarida I added into the link-checker validation for `<script>` tags, and also added validation into the `process-entry-names` plugin to make sure that our bundles are connected correctly. 

I also pulled the vagovprod archive from CI to make sure that our TeamSite bundles are being written correctly.

I ran the CMS Preview Server locally to make sure it works okay, but I don't think there's a good way to test it in production mode, or at least that I know of. We'll find out today at around 3:30 PM, if it suddenly stops working 😬
